File: server/nodejs/src/infrastructure/mappers/mandate-mapper.ts
import { Domain } from 'onlinepayments-sdk-nodejs';
import { Mandate } from '../../business/domain/payments/mandate';
import { RecurrenceType } from '../../business/domain/enums/payments/recurrence-type';
import { CreatePaymentRequestDto } from '../../business/dtos/payment/create-payment/create-payment-request-dto';
import { CountryUtil } from '../../business/utils/models/country-util';

export const MandateMapper = {
    toSdkRequest(requestDto?: CreatePaymentRequestDto | null): Domain.CreateMandateRequest {
        return {
            customer: {
                bankAccountIban: {
                    iban: requestDto?.mandate?.iban,
                },
                mandateAddress: {
                    countryCode: !!requestDto?.mandate?.address?.country
                        ? CountryUtil.toIsoAlpha2(requestDto.mandate.address.country)
                        : undefined,
                    city: requestDto?.mandate?.address?.city,
                    street: requestDto?.mandate?.address?.street,
                    zip: requestDto?.mandate?.address?.zip,
                },
                personalInformation: {
                    name: {
                        firstName: requestDto?.mandate?.address?.firstName,
                        surname: requestDto?.mandate?.address?.lastName,
                    },
                },
            },
            customerReference: requestDto?.mandate?.customerReference,
            recurrenceType: requestDto?.mandate?.recurrenceType?.toString() ?? undefined,
            returnUrl: requestDto?.mandate?.returnUrl,
            signatureType: requestDto?.mandate?.signatureType?.toString() ?? undefined,
        };
    },

    fromSdkResponse(mandate?: Domain.MandateResponse | null): Mandate {
        const recurrenceTypeString = mandate?.recurrenceType ?? null;
        const recurrenceType =
            !!recurrenceTypeString &&
            Object.values(RecurrenceType).includes(recurrenceTypeString.toUpperCase() as RecurrenceType)
                ? (recurrenceTypeString.toUpperCase() as RecurrenceType)
                : undefined;

        return {
            customerReference: mandate?.customerReference,
            iban: mandate?.customer?.bankAccountIban?.iban,
            recurrenceType: recurrenceType,
            mandateReference: mandate?.uniqueMandateReference,
            signatureType: undefined,
            returnUrl: undefined,
            address: {
                lastName: mandate?.customer?.personalInformation?.name?.surname,
                firstName: mandate?.customer?.personalInformation?.name?.firstName,
                city: mandate?.customer?.mandateAddress?.city,
                street: mandate?.customer?.mandateAddress?.street,
                zip: mandate?.customer?.mandateAddress?.zip,
                country: undefined,
            },
        };
    },
};

File: server/python/app/infrastructure/mappers/mandate_mapper.py
from onlinepayments.sdk.domain.create_mandate_request import CreateMandateRequest
from onlinepayments.sdk.domain.mandate_customer import MandateCustomer
from onlinepayments.sdk.domain.bank_account_iban import BankAccountIban
from onlinepayments.sdk.domain.mandate_address import MandateAddress
from onlinepayments.sdk.domain.mandate_personal_information import MandatePersonalInformation
from onlinepayments.sdk.domain.mandate_personal_name import MandatePersonalName
from app.application.domain.enums.common.country import Country
from app.application.domain.payments.mandate import Mandate
from app.application.dtos.get_mandate.response_dto import GetMandateResponseDto

class MandateMapper:

    @staticmethod
    def to_sdk_request(request_dto) -> CreateMandateRequest:
        mandate = request_dto.mandate

        create_mandate = CreateMandateRequest()

        customer = MandateCustomer()

        bank_account = BankAccountIban()
        bank_account.iban = mandate.iban if mandate else None
        customer.bank_account_iban = bank_account

        address = MandateAddress()
        addr = mandate.address if mandate else None
        if addr and addr.country:
            address.country_code = Country.to_iso_alpha2(addr.country)
        address.city = addr.city if addr else None
        address.street = addr.street if addr else None
        address.zip = addr.zip if addr else None
        customer.mandate_address = address

        personal_info = MandatePersonalInformation()
        name = MandatePersonalName()
        name.first_name = addr.first_name if addr else None
        name.surname = addr.last_name if addr else None
        personal_info.name = name
        customer.personal_information = personal_info

        create_mandate.customer = customer
        create_mandate.customer_reference = mandate.customer_reference if mandate else None
        create_mandate.recurrence_type = mandate.recurrence_type.value if mandate and mandate.recurrence_type else None
        create_mandate.return_url = mandate.return_url if mandate else None
        create_mandate.signature_type = mandate.signature_type.value if mandate and mandate.signature_type else None

        return create_mandate

    @staticmethod
    def from_sdk_get_response(sdk_response) -> GetMandateResponseDto:
        mandate = getattr(sdk_response, 'mandate', None)

        return GetMandateResponseDto(
            mandate_reference=getattr(mandate, 'unique_mandate_reference', None)
        )

    @staticmethod
    def from_sdk_create_response(sdk_response) -> Mandate:
        mandate = Mandate()
        mandate.iban = sdk_response.mandate.customer.bank_account_iban.iban
        mandate.customer_reference = sdk_response.mandate.customer_reference
        mandate.return_url = sdk_response.merchant_action.redirect_data.redirect_url

        return mandate